### DISCUSSION TITLE
Problem on interaction between BloCs injected with Riverpod

### DISCUSSION THREAD

**Discussion by maironLucasSlz** [rrousselGit/riverpod#2153] [Q&A]
I work in a project builded around the Provider package to make dependency injection, and also use the BloC pattern to make state management. This last month, we decided to try to migrate from Provider to Riverpod (only for dependency injection) and faced some problems. We have a widget called ActionHandler who receives a stream from BloC and does some stuff. When this stuff involves interaction with a BloC from another widget, this ActionHandler stops work after the first event.

In our case, we have a component to do increments that has its own BloC. On his ActionHandler, has a method call who interacts with the BloC of the page where it is nested. The ActionHandler gets the first event and the method switches the number, but on second click the event is emitted but nothing is received on ActionHandler.

To me, it seems like we are losing reference to the Stream where events flow on the component, but I can't figure out why this is happening, considering that this same logic works well on Provider implementation.

Below, you can see code where page BloC is created:
```
final productDetailBlocProvider = Provider.family
    .autoDispose<ProductDetailBloc, ProductDetailBlocParametersWrapper>(
        (ref, params) {
  final getProductUC = ref.watch(getProductUCProvider);
   (get use case dependencys using Riverpod)
  final bloc = ProductDetailBloc(
    getProductUC: getProductUC,
    productId: params.productId,
  );
  ref.onDispose(bloc.dispose);
  return bloc;
});
```
Below, how we inject BloC on page:
```
  static Widget create({
    String? id,
    String? barcode,
    String? companyId,
    String? selectedOfferId,
  }) =>
      Consumer(
        builder: (context, ref, _) => ProductDetailPage(
          bloc: ref.watch(
            productDetailBlocProvider(
              ProductDetailBlocParametersWrapper(
                companyId: companyId,
                selectedOfferId: selectedOfferId,
                barcode: barcode,
                productId: id,
              ),
            ),
          ),
        ),
      );
```
Then, we pass to the component an anonymous function that adds new value to the BehaviorSubject who handles validations on this value:
```
onValueChanged: (value) {
      bloc.productQuantity.add(value);
},
```
On component, we create and inject BloC in the same way. When a valid value is inserted, we add this to Stream of BloC component that holds ActionEvents:

```
onCounterActionSink.add(
       NotifyValidValue(counterValue: validValue!),
);
```
Then, on Component Widget, we have the ActionHandler, who interacts with the anonymous function passed on Page:
```
return ActionHandler(
  actionStream: widget.bloc.onCounterAction,
  onReceived: (dynamic action) {
    if (action is NotifyValidValue) {
      widget.onValueChanged(action.counterValue);
    }
  },
```
The funny thing is: the first action emits flows as expected, but the second doesn't make it. If I remove this anonymous function call, the ActionHandler receives the events normally, so the problem is in the interaction with another BloC, but I cant figure out why.

**Comment by rrousselGit** [comment]
Hello!

You left out so many things that I can't really help here.
But I'd double-check that ProductDetailBlocParametersWrapper overrides ==. It could also be some issues with your streams.

In any case, the way it is presented, it's not something workable on my end.
If you think you spotted a bug, please make a minimal reproducible example (something which can be executed and which doesn't include your business logic).

In the meantime, I'll convert this to a discussion as it's probably more appropriate

**Reply by maironLucasSlz** [reply]
Hello!
I was working on others issues in my project. Now I am back to Riverpod stuffs. I writed a small example of my problem, it has 140 lines and have same strange behavior than my project. But now, I dont really know if I found a bug or simply dont understand how family modifier works. I need to pass some arguments on BloC creation, reading the doc I think I can do this with family modifier. On doc v2, there are some topics talking about this family parameter be a equatable.
So now I am a little confused, I have a problem with my architecture and how rebuilds works to override BloC or maybe I found a bug when integrating Riverpod with Streams?
Should I put my example here or I need to open a new issue?

**Reply by maironLucasSlz** [reply]
Hello @rrousselGit. Anyway, here is the example you asked. If you could take a look, it would help me a lot (to maybe understand how to use family).
```
void main() {
  runApp(const MyApp());
}

final counterBlocProvider =
    Provider.family.autoDispose<CounterBloc, int>((ref, value) {
  final bloc = CounterBloc(value);
  ref.onDispose(() => bloc.dispose());
  return bloc;
});

final myHomePageBlocProvider = Provider.autoDispose<HomePageBloc>((ref) {
  final bloc = HomePageBloc();
  ref.onDispose(() => bloc.dispose());
  return bloc;
});

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: ProviderScope(
        child: MyHomePage.create(),
      ),
    );
  }
}

class HomePageBloc {
  HomePageBloc();

  final _testSubject = BehaviorSubject<int>.seeded(0);
  Sink<int> get testSink => _testSubject.sink;
  Stream<int> get testStream => _testSubject.stream;

  dispose() {
    _testSubject.close();
  }
}

class CounterBloc {
  CounterBloc(int i) {
    _buttonSubject.listen((_) {
      _actionSubject.add(i++);
    });
  }

  final _actionSubject = BehaviorSubject<int>();
  final _buttonSubject = BehaviorSubject<void>();

  Stream<int> get action => _actionSubject.stream;
  Sink<void> get buttonClick => _buttonSubject.sink;

  dispose() {
    _actionSubject.close();
    _buttonSubject.close();
  }
}

class MyHomePage extends StatelessWidget {
  const MyHomePage({required this.bloc, super.key});

  final HomePageBloc bloc;

  static create() => Consumer(
        builder: (_, ref, __) => MyHomePage(
          bloc: ref.watch(myHomePageBlocProvider),
        ),
      );

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text('Teste'),
      ),
      body: StreamBuilder(
        stream: bloc.testStream,
        builder: (_, snap) {
          return Counter.create(
            (value) => bloc.testSink.add(0),
            snap.data ?? 0,
          );
        },
      ),
    );
  }
}

class Counter extends StatefulWidget {
  const Counter({
    required this.bloc,
    required this.onValueChanged,
    required this.value,
    super.key,
  });

  final CounterBloc bloc;
  final ValueChanged<int> onValueChanged;
  final int value;

  static create(
    ValueChanged<int> onValueChanged,
    int value,
  ) =>
      Consumer(
        builder: (_, ref, __) => Counter(
          onValueChanged: onValueChanged,
          value: value,
          bloc: ref.watch(
            counterBlocProvider(value),
          ),
        ),
      );

  @override
  State<Counter> createState() => _CounterState();
}

class _CounterState extends State<Counter> {
  @override
  void initState() {
    widget.bloc.action.listen((event) {
      print('Event received: $event');
      widget.onValueChanged(event);
    });
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return IconButton(
      onPressed: () => widget.bloc.buttonClick.add(null),
      icon: const Icon(Icons.plus_one),
    );
  }
}
```

**Comment by maironLucasSlz** [comment]
Okay, thank you for replying so quickly!
I will try to make a reproducible example, then I put that here or I should create a new issue?
